---
title: "Convex [Clustering and Biclustering](https://en.wikipedia.org/wiki/Cluster_analysis) with application in R"
author: Marta Karas
date: 30 Sep 2016
output:
  revealjs::revealjs_presentation:
    theme: night
    highlight: espresso
    center: true 
---

<style>
    .reveal h4 {
        text-align: left;
    }
</style>

## My story with <span style="color: orange">`R` in USA</span> 

> * `R` skill brought me to work in USA in Biostatistics Departments<br /><br /> 
> * <span style="color: orange">Top #1 tool</span> used in academic Biostatsistics<br /><br />
> * <span style="color: orange">Reproducibility & usability awareness</span> - rarely one can see a methodology paper published without corresponding `R` package<br /><br /> 



#  <span style="color: orange">Clustering</span> 

## Cluster analysis -  <span style="color: orange">goals</span>   

<br />

> * We perform clustering when we want to <span style="color: orange">group or segment the data set</span> into subsets so that objects within each subset are more closely related to others than those objects assigned to other subsets. 


## Motivation: <span style="color: orange">cancer subtype discovery</span> 

> * A cancer may present clinically as a homogenous disease, but it typically consists of several distinct <span style="color: orange">subtypes at the molecular level</span><br /><br />

> * One of cancer research goals is to identify such cancer subtypes - groups of patients that share <span style="color: orange">distinct genomic signatures and cancer-differing clinical outcomes</span><br /><br />

> * It is the first step towards developing personalized treatment strategies (example: see MD Anderson Cancer Center (Houston, TX) [Personalized Cancer Therapy webpage](https://pct.mdanderson.org/#/) based on Molecular Profiling)



## Cancer subtype discovery as a <span style="color: orange">biclustering</span> problem 

> * We look for subtypes of cancerous tumors that have similar molecular profiles and the genes that characterize each of the them<br /><br />
> * This subtype discovery problem can be posed as a <span style="color: orange">biclustering problem of gene expression
data</span> matrix, where data is partitioned into a checkerboard-like pattern 



## 

![](graphics/biclust-res-example.png) 


## Cancer subtype discovery as a <span style="color: orange">biclustering</span> problem - <span style="color: orange">successess and failures</span>

> * Biclustering breast cancer data has identified sets of genes whose expression levels segregated patients into
five subtypes with distinct survival outcomes (Sørlie et al., 2001)<br /><br />
> * These subtypes have been reproduced in numerous other studies (Sørlie et al., 2003)<br /><br />
> * ... but some other subtypes discoveries were not (i.e. with ovarian cancer (Tothill et al., 2008))


## Cancer subtype discovery as a <span style="color: orange">biclustering</span> problem - <span style="color: orange">challenges</span>

> * The failure to reproduce these other results may reflect an absence of biologically meaningful
groupings<br /><br />
> * ... but another possibility may be related to <span style="color: orange">limitations in the computational
methods currently used</span> to identify biclusters<br /><br />

> * **Goal**: Discover clinically <span style="color: orange">meaningful & reproducible</span> cancer subtypes


# Simple solution: <span style="color: orange">bicluster heatmap</span> with `hclust`

## Simple solution: <span style="color: orange">bicluster heatmap</span> with `hclust`

```{r setup, echo = FALSE, results = "hide", include = FALSE}
# Load all libraries actually here so as to hide packages loading  messages 
library(knitr)
library(s4vd)
library(cvxbiclustr)

opts_knit$set(upload.fun = function(file) imgur_upload(file))
```

```{r, echo = TRUE, results = "hide", fig.height=3.5}
library(s4vd)
data(lung200) # lung cancer data set (Bhattacharjee et al. 2001)

lung200.D.mat <- dist(lung200, method = "euclidean") 
lung200.hclust <- hclust(t(lung200.D.mat), method = "average")

heatmap(lung200, Rowv = as.dendrogram(lung200.hclust), cexRow = 0.3, cexCol = 0.7)
```


## Simple solution: `hclust` - <span style="color: orange">algorithm</span>

1. Begin with $n$ observations and a measure (i.e. Euclidean distance) of all the $n  = n(n − 1)/2$ pairwise dissimilarities. Treat each observation as its own cluster. 

2. For $i=n,n−1,...,2$:

    - Examine all <span style="color: orange">pairwise inter-cluster dissimilarities</span> among the $i$ clusters and identify the pair of clusters that are least dissimilar (that is, most similar). <span style="color: orange">Fuse</span> these two clusters. 
    
    - Compute the new pairwise inter-cluster dissimilarities among the $i-1$ remaining clusters.

## Simple solution: `hclust` - <span style="color: orange">pros and cons</span>

* Pros:

    * Easy to interpret
    * Fast computation 
    
* Cons:

    * <span style="color: orange">Lack of global objective function</span>
    * <span style="color: orange">Instability</span> (subject to perturbations in data, distance choice, linkage type choice)
    * How to choose number of biclusters?
    
<!-- Hierachical clustering cannot be viewed as globally optimizing an objective function -->
<!-- Instead, agglomerative hierarchical clustering techniques use various criteria to decide locally, at each step, which clusters should be merged (or split in divisive approaches) -->


## `hclust` - distance choice effect

```{r, echo = T, fig.width=3, fig.height=4.5, eval = F}
# Function to plot heatmap with hclust dendrogram (method = "average") 
plot.lung200.heatmap.1 <- function(row.dist, col.dist, plt.title){
  heatmap(lung200, main = plt.title,
          cexRow = 0.3, cexCol = 0.7, labRow = FALSE, labCol = FALSE, 
          Rowv = as.dendrogram(hclust(row.dist, method = "average")), 
          Colv = as.dendrogram(hclust(col.dist, method = "average")))
}


D.mat.row <- dist(lung200, method = "euclidean")
D.mat.col <- dist(t(lung200), method = "euclidean")
plot.lung200.heatmap.1(D.mat.row, D.mat.col, "Euclidean dist")

D.mat.row <- dist(lung200, method = "maximum")
D.mat.col <- dist(t(lung200), method = "maximum")
plot.lung200.heatmap.1(D.mat.row, D.mat.col, "Maximum dist")

D.mat.row <-  as.dist(1 - cor(t(lung200)))
D.mat.col <- as.dist(1 - cor(lung200))
plot.lung200.heatmap.1(D.mat.row, D.mat.col, "Correlation dist")
```